### PR TITLE
CO-13 Refactor Dependencies

### PR DESCRIPTION
**Pull request by amirkarimi**
- [x] Refactor `build.sbt` dependencies and moved into `project/CakeDependencies.scala`
- [x] Refactor `project/plugins.sbt` dependencies and moved into `project/project/CakePluginDependencies.scala`
---
- [x] Ready for Review

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -34,44 +34,39 @@ scalacOptions in Compile ++= Seq("-feature", "-deprecation")
...
-addSbtPlugin("com.jsuereth" % "sbt-pgp" % "1.0.1")
-addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.5.15")
-addSbtPlugin("com.typesafe.sbt" % "sbt-digest" % "1.1.1")
-addSbtPlugin("com.typesafe.sbt" % "sbt-git" % "0.9.2")
-addSbtPlugin("com.typesafe.sbt" % "sbt-gzip" % "1.0.0")
-addSbtPlugin("com.typesafe.sbt" % "sbt-native-packager" % "1.1.6")
-addSbtPlugin("io.get-coursier" % "sbt-coursier" % "1.0.0-RC6")
-addSbtPlugin("org.scoverage" % "sbt-scoverage" % "1.5.0")
-addSbtPlugin("org.wartremover" % "sbt-wartremover" % "2.1.0")
+addSbtPlugin(CakeDependencies.Sbt.dynver)
```
Comment (carlpulley) [inline build.sbt]: Lets import `CakeDependencies` at the top of this file?

File: project/CakeDependencies.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
+// Copyright: 2017 https://github.com/cakesolutions/sbt-cake/graphs
+// License: http://www.apache.org/licenses/LICENSE-2.0
+
+import sbt._
+
+object CakeDependencies {
```
Comment (carlpulley) [inline project/CakeDependencies.scala]: For ease of searching, Lets place the top level val's and objects into alphabetical order

File: project/CakeDependencies.scala [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  val leveldbjni = "org.fusesource.leveldbjni" % "leveldbjni" % "1.7"
+  val quasiquotes = "org.scalamacros" %% "quasiquotes" % "2.1.0"
+  val slf4j = "org.slf4j" % "slf4j-api" % "1.7.12"
+
+  object Webjars {
+    val locator = "org.webjars" % "webjars-locator" % "0.26"
+  }
+
+  object Akka {
+    private val version = "2.3.11"
```
Comment (carlpulley) [inline project/CakeDependencies.scala]: Lets not make this private. That way we can potentially use the version value in other code.

**Review by carlpulley — CHANGES_REQUESTED** [review]

**Comment by carlpulley** [comment]
I also note that we have some additional library dependencies in sbt-test - `find . -name "*.sbt" -exec grep -rn % {} \;` should locate the relevant files.

**Comment by carlpulley** [comment]
Won't we also need to refactor library dependencies in CakePlatformPlugin.scala?

**Comment by carlpulley** [comment]
Also, CakeBuildPlugin.scala has a few lines where `"org.scala-lang"` can be replaced by the more generic `scalaOrganization.value`.

A similar comment applies to project/plugins.sbt

File: build.sbt (hunk 1 above)
Comment (amirkarimi) [reply build.sbt]: Then we'll get the following error as `Sbt` is also imported by another plugin:
```
error: reference to Sbt is ambiguous;
it is imported twice in the same scope by
import CakeDependencies._
and import _root_.com.lucidchart.sbt.scalafmt.ScalafmtSbtPlugin.autoImport._
```

**Comment by amirkarimi** [comment]
@carlpulley 
> I also note that we have some additional library dependencies in sbt-test - find . -name "*.sbt" -exec grep -rn % {} \; should locate the relevant files.

All of them reference to `sbt-cake` and they're just single dependencies, except one which references to `akka-http-spray-json` and again is a single dependency. As they're separate projects, it doesn't make sense to me to refactor those single line dependencies into another file for each project. Or is there a better way?

> Won't we also need to refactor library dependencies in CakePlatformPlugin.scala?

`CakePlatformKeys` itself is just a list of dependencies. Though we can move it to a separate file.

**Comment by amirkarimi** [comment]
>Also, CakeBuildPlugin.scala has a few lines where "org.scala-lang" can be replaced by the more generic scalaOrganization.value.
>
>A similar comment applies to project/plugins.sbt

The only place that `org.scala-lang` is used is [here](https://github.com/cakesolutions/sbt-cake/blob/master/src/main/scala/net/cakesolutions/CakeBuildPlugin.scala#L104-L114) which seems to be intentional.

File: project/CakeDependencies.scala [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  object Commons {
+    val codec = "commons-codec" % "commons-codec" % "1.6"
+    val comporess ="org.apache.commons" % "commons-compress" % "1.9"
+    val langs = "org.apache.commons" % "commons-lang3" % "3.4"
+    val logging = "commons-logging" % "commons-logging" % "1.1.3"
+  }
+
+  object Sbt {
```
Comment (carlpulley) [inline project/CakeDependencies.scala]: If we place these `Sbt` dependencies in `CakePluginDependencies`, would we be able to access them in this context/scope?

Comment (amirkarimi) [reply project/CakeDependencies.scala]: As far as I know we can't have access to objects defined in `project` folder (meta-build) in the main code (proper build) as we can't obtain the project name in the main code directly from `build.sbt`. If it's true, the same rule applies here; the context of `/project/project/CakePluginDependencies.scala` is different than the files in `/project`.

More info: 
- http://www.scala-sbt.org/0.13/docs/Organizing-Build.html
- https://stackoverflow.com/questions/23944108/how-to-share-version-values-between-project-plugins-sbt-and-project-build-scala

**Comment by amirkarimi** [comment]
After having a discussion with @carlpulley we decided to keep the current PR at this state and create a new ticket to solve the issue: https://cakesolutions.jira.com/browse/CO-143

File: build.sbt [hunk 5]
```diff
@@ -34,29 +35,29 @@ scalacOptions in Compile ++= Seq("-feature", "-deprecation")
...
+addSbtPlugin(SbtDependencies.scalafmt)
 
 enablePlugins(ScalafmtPlugin)
 
 // TODO: CO-68: remove JSR305 dependency when SBT moves away from Scala 2.10
-val findbugs = "com.google.code.findbugs" % "jsr305" % "3.0.2"
-libraryDependencies += findbugs % "provided"
+libraryDependencies += jsr305 % "provided"
 
+// These are dependency overrides which are different than the regular dependencies and should be maintained here. Note: Ideally we should refactor all dependencies in a single place. Read more: https://cakesolutions.jira.com/browse/CO-143
```
Comment (carlpulley) [inline build.sbt]: This line length is too long

File: project/CakeDependencies.scala [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    val packager = "com.typesafe.sbt" % "sbt-native-packager" % "1.1.6"
+    val pgp = "com.jsuereth" % "sbt-pgp" % "1.0.1"
+    val plugin = "com.typesafe.play" % "sbt-plugin" % "2.5.15"
+    val scalafmt = "com.lucidchart" % "sbt-scalafmt-coursier" % "1.7"
+    val scoverage = "org.scoverage" % "sbt-scoverage" % "1.5.0"
+    val wartRemover = "org.wartremover" % "sbt-wartremover" % "2.1.0"
+    val web = "com.typesafe.sbt" % "sbt-web" % "1.3.0"
+  }
+}
+
```
Comment (carlpulley) [inline project/CakeDependencies.scala]: Delete extra line please

File: src/main/scala/net/cakesolutions/CakePlatformPlugin.scala [hunk 7]
```diff
@@ -15,6 +15,7 @@ import wartremover._
 /**
   * Library dependency keys that will be auto-imported when this plugin is
   * enabled on a project.
+  * Note: Ideally we should refactor all dependencies in a single place. Read more: https://cakesolutions.jira.com/browse/CO-143
```
Comment (carlpulley) [inline src/main/scala/net/cakesolutions/CakePlatformPlugin.scala]: Let's make this a TODO: with just the ticket number?

File: project/CakeDependencies.scala [hunk 8]
```diff
@@ -0,0 +1,33 @@
+// Copyright: 2017 https://github.com/cakesolutions/sbt-cake/graphs
+// License: http://www.apache.org/licenses/LICENSE-2.0
+
+import sbt._
+
+/**
+  * Dependencies referenced in sbt-cake build.sbt
+  * Note: Ideally we should refactor all dependencies in a single place. Read more: https://cakesolutions.jira.com/browse/CO-143
```
Comment (carlpulley) [inline project/CakeDependencies.scala]: Please make a TODO

File: project/project/CakePluginDependencies.scala [hunk 9]
```diff
@@ -0,0 +1,21 @@
+// Copyright: 2017 https://github.com/cakesolutions/sbt-cake/graphs
+// License: http://www.apache.org/licenses/LICENSE-2.0
+
+import sbt._
+import sbt.Keys._
+
+/**
+  * Dependencies referenced in sbt-cake plugins
+  * Note: Ideally we should refactor all dependencies in a single place. Read more: https://cakesolutions.jira.com/browse/CO-143
```
Comment (carlpulley) [inline project/project/CakePluginDependencies.scala]: Please make a TODO

**Review by carlpulley — CHANGES_REQUESTED** [review]

**Review by carlpulley — APPROVED** [review]

**Review by ferhtaydn — APPROVED** [review]

**Pull request merged**